### ISSUE TITLE
Wrong default mode for mkdir?

### ISSUE THREAD

**Issue by mmenozzi** [amphp/file#25]
Hi,
the `Amp\File\mkdir` function defines `0644` as default mode (https://github.com/amphp/file/blob/master/lib/functions.php#L245).
But PHP's `\mkdir` function defines the same default mode as `0777` (http://php.net/manual/en/function.mkdir.php).

Is there any reason why the default mode of `Amp\File` is different from the PHP one?

**Comment by kelunik** [comment]
Yes, there is, security. It's very bad for every file to be 777 by default. But I think there's still the default `umask` applied.

**Comment by bwoebi** [comment]
Yeah, the umask is always applied. So actually shouldn't matter.

**Comment by mmenozzi** [comment]
The umask is applied making a bitwise AND with the mode. On most system the default umask is 0022, which applied to this default mode 0644 produces a 644 permission. This permission is wrong for directories because it lacks the execution flag.

**Comment by kelunik** [comment]
Indeed, should be 755 for those.

**Comment by mmenozzi** [comment]
Indeed 0777 & 0022 = 0755

**Comment by kelunik** [comment]
Do you want to provide a PR?

**Comment by mmenozzi** [comment]
I mean the default mode of Amp\File of 0644 with default umask 0022 produces 0644 (wrong). If it were 0777 (like php's mkdir) it would produce 0755 (proper dir permission).

**Comment by mmenozzi** [comment]
Yes.

**Comment by mmenozzi** [comment]
Yes I'll provide a PR (sorry for multiple comments I'm on mobile)

**Comment by kelunik** [comment]
@mmenozzi I'm not sure whether UV and EIO respect the mask, we have to test that.

**Comment by mmenozzi** [comment]
Ok @keluinik

**Comment by mmenozzi** [comment]
Hi @kelunik,
on my dev machine with PHP 7.1.10 I installed `uv` and `eio` extensions. Then I ran the following script:

```php
uv_fs_mkdir(uv_default_loop(), '/tmp/testuv', 0777);
eio_mkdir('/tmp/test', 0777);
```

This is the result (umask is `0022` as said before):

```
drwxr-xr-x   2 manuele  wheel   64 Mar 15 09:26 test
drwxr-xr-x   2 manuele  wheel   64 Mar 15 09:38 testuv
```

So both `eio` and `uv` extensions work like the core PHP `mkdir` function applying the umask.
Can we proceed with the PR after this? Or maybe we should add a specific unit test about this case in `Amp\File`?
